java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/SlidingWindowUtils.java

```diff
@@ -469,6 +456,110 @@ public static Row updateZeroValuedRow(
+     *
+     * @param tEnv The StreamTableEnvironment of the table.
+     * @param table The input table.
+     * @param windowDescriptor The descriptor of the sliding window to be applied to the input *
+     *     table.
+     * @param aggDescriptors The descriptor of the aggregation field in the sliding * window.
+     * @param rowTimeFieldName The name of the row time field.
+     * @return The result table.
+     */
+    public static Table applyGlobalWindow(
+            StreamTableEnvironment tEnv,
+            Table table,
+            SlidingWindowDescriptor windowDescriptor,
+            AggregationFieldsDescriptor aggDescriptors,
+            String rowTimeFieldName) {
+        if (windowDescriptor.stepSize != Duration.ZERO
+                || aggDescriptors.getMaxWindowSizeMs() != 0) {
+            throw new RuntimeException("Step size and window size must be 0.");
+        }
+
+        table = populateFilterExprFlag(table, aggDescriptors);
+
+        ResolvedSchema schema = table.getResolvedSchema();
+
+        DataStream<Row> stream =
+                tEnv.toChangelogStream(
+                        table,
+                        Schema.newBuilder().fromResolvedSchema(schema).build(),
+                        ChangelogMode.insertOnly());
+
+        List<String> fieldNames = new ArrayList<>();
+        List<TypeInformation<?>> accumulatorFieldTypes = new ArrayList<>();
+        final Map<String, DataType> resultDataTypeMap = new HashMap<>();
+        for (String key : windowDescriptor.groupByKeys) {
+            fieldNames.add(key);
+            accumulatorFieldTypes.add(ExternalTypeInfo.of(getDataType(schema, key).nullable()));
+            resultDataTypeMap.put(key, getDataType(schema, key).nullable());
+        }
+        fieldNames.add(rowTimeFieldName);
+        accumulatorFieldTypes.add(ExternalTypeInfo.of(DataTypes.TIMESTAMP_LTZ(3).nullable()));
+        resultDataTypeMap.put(rowTimeFieldName, DataTypes.TIMESTAMP_LTZ(3).nullable());
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggDescriptor :
+                aggDescriptors.getAggFieldDescriptors()) {
+            fieldNames.add(aggDescriptor.fieldName);
+            accumulatorFieldTypes.add(
+                    aggDescriptor.aggFuncWithoutRetract.getAccumulatorTypeInformation());
+            resultDataTypeMap.put(aggDescriptor.fieldName, aggDescriptor.dataType);
+        }
+
+        final String[] keyFieldNames = windowDescriptor.groupByKeys.toArray(new String[0]);
+        final List<DataTypes.Field> resultTableFields =
+                getResultTableFields(
+                        resultDataTypeMap, aggDescriptors, rowTimeFieldName, keyFieldNames);
+        final ExternalTypeInfo<Row> resultRowTypeInfo =
+                ExternalTypeInfo.of(
+                        DataTypes.ROW(resultTableFields.toArray(new DataTypes.Field[0])));
+
+        SingleOutputStreamOperator<Row> resultStream =
```

Comment:
The computation logic above this line seems very similar to that of `applySlidingWindowPreAggregationProcess`. Can we introduce a private method to deduplicate the common part?

Reply:
It turns out that the accumulator-type information is unnecessary so I refactor the code.